lib/elixir/test/elixir/kernel/binary_test.exs

```diff
@@ -244,6 +244,16 @@ defmodule Kernel.BinaryTest do
     assert (fn <<_, _::size(x)>> -> true end).(<<?a, ?b>>)
   end
 
+  test "bitsyntax size using guard expressions" do
+    x = 8
+    assert <<1::3>> == <<1::size(x - 5)>>
+    assert <<1::3*8>> == <<1::size(x - 5)-unit(8)>>
+    assert <<1::4>> == <<1::size(length('abcd'))>>
+
+    foo = %{bar: 5}
+    assert <<1::5>> == <<1::size(foo.bar)>>
```

Comment:
The guard expressions should only be checked on a match, so you need to rewrite to this:

```suggestion
    x = 8
    assert <<1::size(x - 5)>> = <<1::3>>
    assert <<1::size(x - 5)-unit(8)>> = <<1::3*8>>
    assert <<1::size(length('abcd'))>> = <<1::4>>

    foo = %{bar: 5}
    assert <<1::size(foo.bar)>> = <<1::5>>
```